### ISSUE TITLE
(Presumably) incomplete processing of video

### ISSUE THREAD

**Issue by Ferdinandinio** [Breakthrough/DVR-Scan#54]
Hello, great software for finding the birds on my birdcam. 

However, it seems that the processing of the video files never reaches 70% of total frames.
The input files are .AVI and no matter the output format, this issue persists. I don't know if it is just a visual problem or if the files are not completely processed.

In-/Output below:

```
$ dvr-scan -i VIDEO.avi -roi 697 400 536 452 -l 30 -o MO_VIDEO.mp4
OpenCV: FFMPEG: tag 0x44495658/'XVID' is not supported with codec id 12 and format 'mp4 / MP4 (MPEG-4 Part 14)'
OpenCV: FFMPEG: fallback to use tag 0x7634706d/'mp4v'
[DVR-Scan] Initializing scan context...
[DVR-Scan] Opened video VIDEO.avi (1920 x 1080 at 30.000 FPS).
[DVR-Scan] Appended video VIDEO.avi.
[DVR-Scan] ROI selected (x,y,w,h): [697, 400, 536, 452]
[DVR-Scan] Processed:   0%|          | 0/17999 [00:00<?, ? frames/s][DVR-Scan] Scanning input video for motion events...
[DVR-Scan] Processed:  69%|######8   | 12388/17999 [02:49<01:16, 72.94 frames/s][DVR-Scan] Processed 12388 frames read in 169.8 secs (avg 72.9 FPS).
[DVR-Scan] Detected 1 motion events in input.
[DVR-Scan] Scan-only mode specified, list of motion events:
-------------------------------------------------------------
|   Event #    |  Start Time  |   Duration   |   End Time   |
-------------------------------------------------------------
|  Event    1  |  00:00:07.6  |  00:00:03.2  |  00:00:10.7  |
-------------------------------------------------------------
[DVR-Scan] Comma-separated timecode values:
00:00:07.567,00:00:10.733
[DVR-Scan] Motion events written to disk.
```

```
$ dvr-scan -i VIDEO.avi -roi 697 400 536 452 -l 30 -o MO_VIDEO.avi
[DVR-Scan] Initializing scan context...
[DVR-Scan] Opened video VIDEO.avi (1920 x 1080 at 30.000 FPS).
[DVR-Scan] Appended video VIDEO.avi.
[DVR-Scan] ROI selected (x,y,w,h): [697, 400, 536, 452]
[DVR-Scan] Processed:   0%|          | 0/17999 [00:00<?, ? frames/s][DVR-Scan] Scanning input video for motion events...
[DVR-Scan] Processed:  69%|######8   | 12377/17999 [04:11<01:54, 49.17 frames/s][DVR-Scan] Processed 12377 frames read in 251.7 secs (avg 49.2 FPS).
[DVR-Scan] No motion events detected in input.
```

**Comment by Breakthrough** [comment]
Hi @Ferdinandinio;

To assist with the investigation, could you provide some additional details:

 1. How large are the files?  Are you able to upload one to this issue?
 2. What is the reported length of the video when you play it back locally (e.g. using VLC)?  Can you seek past 7 minutes?
 3. Please run the video through `ffprobe` and provide the output.  (e.g. run `$ ffprobe MO_VIDEO.mp4`)
 4. Could you try converting the video to .mkv (either with ffmpeg or mkvmerge) and see if the output differs?  Could you also share how you converted the mp4 to AVI?

Thanks!

**Comment by PatrikKnizka** [comment]
Hey, I have the same problem with my files. 

What I want to do with DVR-Scan is to scan through 24h footage of security camera and look for motion on one exact spot. Everything works well until 42% of the whole scan. 

Suddenly the process stops and it looks like it has already ended, but it want through only 42% of the existing frames (for example it scanned me only 45 200/108 697 frames). The program doesn't show any error or anything.

And the answers to your questions above:
1. For me the same problem is with all files, so I think the size of file does not matter. 
But lets take an example file.. footage "00-01.avi" has 108MB

2. Reported length of the video after opening in VLC and also the length shown in Windows file explorer is the same - 1:00:00
I can watch the whole footage, but when I want to skip some part the VLC has some problems. The video looks like there is lag and then VLC crashes from no reason.

3. Output from ffprobe of "00-01.avi" is here:
`Input #0, avi, from '00-01.avi':
  Duration: 01:00:00.63, start: 0.000000, bitrate: 253 kb/s
  Stream #0:0: Video: hevc (Main) (H265 / 0x35363248), yuvj420p(pc, bt709), 1920x1080, 247 kb/s, 30 fps, 30 tbr, 30 tbn`

4. If it is necessary I can try it, but 1 hour footage would take a lot time to be converted into another type of file.

**Comment by Breakthrough** [comment]
@PatrikKnizka if you use mkvmerge it shouldn't take too long.  If you can't view the video with VLC after seeking, however, that indicates something is wrong with the video.  While sometimes it works, I'm using OpenCV for the video I/O, which hides a lot of the video decoding logic.  I can consider other backends though, like PyAV, as I'm planning for PySceneDetect.

Try using `mkvmerge` and see if that makes a difference.  If not, then you can use the following command to reencode your videos fast without quality loss (although larger file size):

    ffmpeg -i input.mp4 -c:v libx264 -preset fast -crf 0 -c:a copy output.mp4

If you could try that as well that would be a big help, or if you can share a sample file somehow.  Thanks!

**Comment by PatrikKnizka** [comment]
Yes, I also thought there is some problem with the video, since the VLC had some problems with it, but I don't know what kind of problem is it, because the video footage is downloaded right from our camera system at home. When I want to watch the same footage right on camera system connected to the TV it works perfectly fine. 

Maybe our camera system is too old so it's using some old software. Nevermind.

- I tried `mkvmerge -o output1.mkv 00-01.avi`, the process took just few seconds. Here is the output from CLI. 
`C:\Users\Patrik\Desktop\05082021>mkvmerge -o output2.mkv 00-01.avi
mkvmerge v60.0.0 ('Are We Copies?') 64-bit
'00-01.avi': Using the demultiplexer for the format 'AVI'.
'00-01.avi' track 0: Using the output module for the format 'VfW compatible video'.
The file 'output2.mkv' has been opened for writing.
Progress: 100%
The cue entries (the index) are being written...
Multiplexing took 3 seconds.`
But problem is, that also after mkvmerge I can not skip parts in VLC (same problem as described above). So I think the problem persists.

- I also tried `ffmpeg -i 00-01.avi -c:v libx264 -preset fast -crf 0 -c:a copy output.mp4, the process took few minutes. Here is the output from CLI:
`Input #0, avi, from '00-01.avi':
  Duration: 01:00:00.63, start: 0.000000, bitrate: 253 kb/s
  Stream #0:0: Video: hevc (Main) (H265 / 0x35363248), yuvj420p(pc, bt709), 1920x1080, 247 kb/s, 30 fps, 30 tbr, 30 tbn
Stream mapping:
  Stream #0:0 -> #0:0 (hevc (native) -> h264 (libx264))
Press [q] to stop, [?] for help
[libx264 @ 000001f21454eac0] using cpu capabilities: MMX2 SSE2Fast SSSE3 SSE4.2 AVX FMA3 BMI2 AVX2
[libx264 @ 000001f21454eac0] profile High 4:4:4 Predictive, level 4.0, 4:2:0, 8-bit
[libx264 @ 000001f21454eac0] 264 - core 164 r3065 ae03d92 - H.264/MPEG-4 AVC codec - Copyleft 2003-2021 - http://www.videolan.org/x264.html - options: cabac=1 ref=2 deblock=1:0:0 analyse=0x3:0x113 me=hex subme=6 psy=0 mixed_ref=1 me_range=16 chroma_me=1 trellis=0 8x8dct=1 cqm=0 deadzone=21,11 fast_pskip=0 chroma_qp_offset=0 threads=12 lookahead_threads=2 sliced_threads=0 nr=0 decimate=1 interlaced=0 bluray_compat=0 constrained_intra=0 bframes=0 weightp=1 keyint=250 keyint_min=25 scenecut=40 intra_refresh=0 rc=cqp mbtree=0 qp=0
Output #0, mp4, to 'output.mp4':
  Metadata:
    encoder         : Lavf59.4.101
  Stream #0:0: Video: h264 (avc1 / 0x31637661), yuvj420p(pc, bt709, progressive), 1920x1080, q=2-31, 30 fps, 15360 tbn
    Metadata:
      encoder         : Lavc59.4.101 libx264
    Side data:
      cpb: bitrate max/min/avg: 0/0/0 buffer size: 0 vbv_delay: N/A
More than 1000 frames duplicated  81408kB time=00:00:56.36 bitrate=11831.4kbits/s dup=995 drop=0 speed=3.74x
More than 10000 frames duplicated817920kB time=00:09:29.06 bitrate=11774.4kbits/s dup=9980 drop=0 speed=3.53x
frame=108019 fps= 96 q=-1.0 Lsize= 5194958kB time=01:00:00.60 bitrate=11819.4kbits/s dup=63099 drop=0 speed=3.21x
video:5194423kB audio:0kB subtitle:0kB other streams:0kB global headers:0kB muxing overhead: 0.010310%
[libx264 @ 000001f21454eac0] frame I:433   Avg QP: 0.00  size:620661
[libx264 @ 000001f21454eac0] frame P:107586 Avg QP: 0.00  size: 46942
[libx264 @ 000001f21454eac0] mb I  I16..4: 31.1%  7.9% 61.0%
[libx264 @ 000001f21454eac0] mb P  I16..4:  0.5%  0.5%  0.5%  P16..4: 11.6%  0.6%  0.5%  0.0%  0.0%    skip:85.8%
[libx264 @ 000001f21454eac0] 8x8 transform intra:26.9% inter:64.9%
[libx264 @ 000001f21454eac0] coded y,uvDC,uvAC intra: 98.8% 0.0% 0.0% inter: 12.0% 0.0% 0.0%
[libx264 @ 000001f21454eac0] i16 v,h,dc,p: 26% 74%  0%  0%
[libx264 @ 000001f21454eac0] i8 v,h,dc,ddl,ddr,vr,hd,vl,hu: 37% 62%  0%  0%  0%  0%  0%  0%  0%
[libx264 @ 000001f21454eac0] i4 v,h,dc,ddl,ddr,vr,hd,vl,hu: 44% 44%  2%  1%  3%  2%  2%  1%  1%
[libx264 @ 000001f21454eac0] i8c dc,h,v,p: 100%  0%  0%  0%
[libx264 @ 000001f21454eac0] Weighted P-Frames: Y:0.0% UV:0.0%
[libx264 @ 000001f21454eac0] ref P L0: 99.6%  0.4%
[libx264 @ 000001f21454eac0] kb/s:11818.12`
The output.mp4 video is playable and works perfectly fine in VLC. No problem with skipping parts or anything else. So I am gonna try out to run dvr-scan one more time with this file.

**Comment by PatrikKnizka** [comment]
After trying `dvr-scan -i output.mp4 -o ouput_test.mp4 -roi 1415 471 113 97` some problem occurs:
`OpenCV: FFMPEG: tag 0x44495658/'XVID' is not supported with codec id 12 and format 'mp4 / MP4 (MPEG-4 Part 14)'
OpenCV: FFMPEG: fallback to use tag 0x7634706d/'mp4v'`

But it did not stop the process so dvr-scan continued to scan through the video. (The amount of all frames is approximately the same as before - 108 019)
The process was also not stopped on 42% as before and continued to 100%.

**The result:**
`[DVR-Scan] Processed: 100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 108019/108019 [11:42<00:00, 153.75 frames/s]
[DVR-Scan] Processed 108019 frames read in 703.4 secs (avg 153.6 FPS).
[DVR-Scan] Detected 49 motion events in input.`
It seems like it went through the whole footage.
The output_test.mp4 video works perfectly fine and it is playable not only in VLC but also in Windows player. So I guess that's the solution for our problem.

**Comment by Breakthrough** [comment]
This may likely be solved by #62.  Feel free to let me know if the latest version of DVR-Scan still encounters the issue.
